src/Symfony/Component/Form/Form.php

```diff
@@ -795,6 +795,25 @@ public function getErrors($deep = false, $flatten = true)
         return new FormErrorIterator($this, $errors);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function clearErrors($deep = false)
+    {
+        $this->errors = array();
+
+        if ($deep) {
+            // Clear errors from children
+            foreach ($this as $child) {
+                if ($child instanceof ClearableErrorInterface) {
+                    $child->clearErrors(true);
```

Comment:
Is it possible to have a clearable form type nested in a non-clearable one?

Reply:
Yes, buttons are not clearable and they could be included in a form that is clearable.

Reply:
The question was the other way round. As the code here stops the traversal of the form tree as soon as a non-clearable form type is encountered, it will not clear clearable fields nested in non-clearable ones.

Reply:
The only non-clearable fields we have in the symfony core are buttons that can't have children.

This could append in libraries but then it can be considered the responsibility of the library to implement this interface if needed.

Reply:
Yeah, I know. But why risk it, if the change would be fairly easy?

Reply:
It would mean children won't be able to implement custom error cleaning logic, that's something someone might want to do.

Reply:
> Yeah, I know. But why risk it, if the change would be fairly easy?

We can't add `clearErrors()` to `FormInterface` as that would be a BC break.  And even if we could, that would require us to implement `clearErrors()` on the `Button` class, which makes no sense because buttons can't have errors or children - it'll either be a pointless no-op or throw a `BadMethodCallException`.

Reply:
I wasn't hinting at implementing it everywhere, but decoupling the traversal of the form tree from the actual form field clearing.

Like the visitor pattern, if you will.
But this edge case doesn't seem too severe.